src/Parser.php

```diff
@@ -317,7 +319,7 @@ protected function filterHost($host)
      */
     protected function isIpv6Host(string $ipv6): bool
     {
-        if ('][' !== substr($ipv6.$ipv6, strlen($ipv6) - 1, 2)) {
+        if (2 > strlen($ipv6) || '[' !== $ipv6[0] || ']' !== substr($ipv6, -1)) {
```

Comment:
the `$ipv6` string can not be the empty string.. the check is done prior to calling this method so you don't need to check its length.

Reply:
Since the method is protected, technically no assumptions should be made about the input here.

Reply:
yes the method is protected (should have been private .. and the whole class final ... but that's another topic for another day)  meaning as a maintainer I do know when and why it will be used so I'm in the camp that I can make assumptions. If it was a public API method I'll second your change because there I have no way to control the user input.

Reply:
Instead of a function call to `strlen`, this could use `($ipv6[0] ?? '') !== ...`. On the other hand, `\strlen` is optimized to its own opcode anyway, but only if it's prefixed with `\`.